### ISSUE TITLE
Massive performance decrease since 3.3.x

### ISSUE THREAD

**Issue by bluemoehre** [audacity/audacity#4764]
### Bug description

I am working since a few years on a huge mix to challenge myself using only Audacity. This project already has >30 tracks and a total playtime >40mins. It mostly consists of hard cuts and rarely has effects etc.
It already became slower and slower with the past versions, but still it was okayish to work with. Now I upgraded to 3.3.x and performance was hit very hard.

In parallel I set up a new machine and the hardware is now 130% on single and 400% faster on multicore. Also switched from 16GB to 32GB Ram and migrated all drives to M.2. But this new machine didnt made any change. Also I can see the process to take like <10% load, low i/o, nothing special - system is bored I would say ;)

But the responsiveness of the UI is like watching a movie with 2fps. While scrolling and navigating the timeline there is a huuuge bottleneck somwhere. Lets say: it iss no more usable.

### Steps to reproduce

1. Create a project 44.100Hz, 32bit
2. add >30 FLAC songs
3. just arrange them one after each other but each in its own track

### Expected behavior

I can work with the project flawlessly

### Actual behavior

UI performance is so bad, mouse and keyboard interactions get queued for multiple seconds

### Audacity Version

latest stable version (from audacityteam.org/download)

### Operating system

Windows

### Additional context

i5-13600k
32GB DDR5
3x 1TB WD Black m2 SSD
ASUS Essence STX II
AMD RX480 @ 3840x2160 60Hz

**Comment by crsib** [comment]
Can you please install any [nightly](https://nightly.link/audacity/audacity/workflows/build/master)? There is a Frame Statistics tool under Help -> Diagnostics, I wonder what it shows you.

**Comment by bluemoehre** [comment]
I can do that. In the meanwhile to let you know: it seems to be kinda memory leak. Performance keeps decreasing while arranging audio. After a save and restart it is getting way better - for lets say 30mins, then its unusable

**Comment by bluemoehre** [comment]
Hope this helps finding the issue. This is after 1min ...

![Screenshot 2023-06-02 235824](https://github.com/audacity/audacity/assets/1735386/8dd6263a-6782-4a55-9576-a547d33b178f)

**Comment by dozzzzer** [comment]
Couldn't reproduce the issue on either of my machines: Macbook M1 Pro, Win 11 VM, and the AMD Ryzen 4700U Win 10 laptop.
This is as bad as it gets for me when scrolling, and zooming in/out while playing 30+ flac files arranged one after another, each on a separate track, which isn't perfect but is far from unusable:

![image](https://github.com/audacity/audacity/assets/70063970/b9e24c9d-6c40-4928-9a18-e8b209710d8b)

Seems like some crucial steps are missing here, @bluemoehre

**Comment by bluemoehre** [comment]
Do you also have a Windows machine to check if it's a OS related issue @dozzzzer ?

**Comment by crsib** [comment]
Well, it indeed depends on the Audacity window size, but with the 2560x1440 I can only get as bad as 150 msec for max. I have 5800x which appears to have 25 to 30% higher single-core performance, but it still does not explain 1-sec max and 200 msec avg you observe.

> Any idea how to get more debugging details, what is causing this?

It would help if you can provide us with some detailed steps (like opening Audacity, importing a file, applying the following effect, etc) or even a video of your process. Given that restarting helps you - probably there is something in your project history that causes trouble or the problem is related to how often your project is saved to the recovery table.

**Comment by bluemoehre** [comment]
I reinvestigated a bit keeping the Frame Stats visible. 

I can clearly see a **performance decrease** by:

- **total tracks in project** (linear, ~2ms each track)
- **total visible tracks** (linear, ~10ms each track)
- **total visible waveform blocks** (unknown, <10 linear, >10 no change)
- **current edit mode** (~20ms Multi-Edit, )
- **application window size** (linear, related to above factors I guess)
- Resolution / **DPI** (linear, related to above factors I guess)
- Application **runtime** (unknown factor)

As said **I am using 4K high res display** here. Playing around with **resolution** and system **UI scaling**, I can also see performance changes. (Unfortunaltely, High DPI scaling is not supported in Audacity since ever, so I mostly use the System's Upscaler - but that's another topic and not directly related I guess.)

When rendering the UI in native **8K** (via graphics driver upscaler) I can see **linear performance decrease** up to **5000ms** frame time 🐌.

So I assume something in the track render logic is inperformant and/or there is some issue with cache access for waveforms. 
Additionally I want to point out again the age of project. It was initially started years ago - maybe that's an issue as well?


Here is a screenshot how the project looks like:

![Screenshot 2023-06-15 112528](https://github.com/audacity/audacity/assets/1735386/e36cae4c-229c-4328-bbea-1a01c3de6c3b)

**Comment by crsib** [comment]
> track render logic is inperformant

It is indeed, but it has stayed the same for 15+ years. An attempt to improve the sitaution resulted in a very complex 25k+ lines pull request, which made itself unmergeable almost immediately. This issue will be handled very carefully in Audacity 4.  

> It was initially started years ago - maybe that's an issue as well?

That is an excellent question. Can you please run https://github.com/audacity/audacity-project-tools as `audacity-project-tools -analyze_project project.aup3` and share your output?

**Comment by crsib** [comment]
The track height was the case. I've managed to get 500msec locally.

**Comment by bluemoehre** [comment]
Okay, so I guess we have the main issue. Anyhow I checked the SQLite and did not see any anomalies in the the output. But the last lines could be interesting:

```
[ ... ]
Track 207: DJ Gollum feat. Scarlett - Poison (Marco Van Bassken Radio Edit)
        Clip 0: ''.
                Total samples 5111769
                Total time: 01:55.913
                Trimmed time: 01:55.913
                Trimmed / Total: 100.0000%
Total blocks in project: 2002
        Silent blocks count: 148 (7.3926%)
Not shared blocks count: 1880 (93.906%)
        Silent blocks count: 96 (5.1064%)
```

Looks like I underestimated the amount of track by far 🙊

**Comment by bluemoehre** [comment]
Is Audacity reading the `.aup3` and waveforms directly from the drive and not from a memory cache? It looks like disabling the Windows File Compression improved the performance by 15%. 
(This compression gets automatically enabled again and again when drive gets filled >80%. Besides that I never thought it would make any difference on m.2 having boored CPU cores hanging around)

**Comment by crsib** [comment]
> waveforms directly from the drive and not from a memory cache

There is very little caching when drawing waveforms ATM. I will check how it runs on HDD out of curiosity (and for fun!).

(And just for fun also - here is a build from the abandoned branch that improves performance
[audacity-rendering-branch.zip](https://github.com/audacity/audacity/files/11770541/audacity-rendering-branch.zip))

**Comment by LWinterberg** [comment]
I'll close this issue as it's not really actionable - as Dmitry said, this will be taken care of for Audacity 4, and can't really be fixed before that.

**Comment by bluemoehre** [comment]
Hey @LWinterberg @crsib ,

the issue seems to be fixed in v3.4.2. At least the UI feels more snappy; scrolling/zooming is way more fast. Subjectively I feel an improvement from 5fps to 30fps. 

Many thanks!
